config/config.go

```diff
@@ -1,269 +1,359 @@
+	c.DefaultedGlobalConfig = DefaultGlobalConfig
+	if err := unmarshal(&c.DefaultedGlobalConfig); err != nil {
+		return err
 	}
-	return duration
-}
-
-// ScrapeInterval gets the default scrape interval for a Config.
-func (c *Config) ScrapeInterval() time.Duration {
-	return stringToDuration(c.Global.GetScrapeInterval())
+	return nil
 }
 
-// EvaluationInterval gets the default evaluation interval for a Config.
-func (c *Config) EvaluationInterval() time.Duration {
-	return stringToDuration(c.Global.GetEvaluationInterval())
+// DefaultedGlobalConfig is a proxy type for GlobalConfig.
+type DefaultedGlobalConfig struct {
+	// How frequently to scrape targets by default.
+	ScrapeInterval Duration `yaml:"scrape_interval"`
+	// The default timeout when scraping targets.
+	ScrapeTimeout Duration `yaml:"scrape_timeout"`
+	// How frequently to evaluate rules by default.
+	EvaluationInterval Duration `yaml:"evaluation_interval"`
+
+	// The labels to add to any timeseries that this Prometheus instance scrapes.
+	Labels clientmodel.LabelSet `yaml:"labels,omitempty"`
 }
 
-// ScrapeConfig encapsulates a protobuf scrape configuration.
+// ScrapeConfig configures a scraping unit for Prometheus.
 type ScrapeConfig struct {
-	pb.ScrapeConfig
+	// DefaultedScrapeConfig contains the actual fields for ScrapeConfig.
+	DefaultedScrapeConfig `yaml:",inline"`
 }
 
-// ScrapeInterval gets the scrape interval for the scrape config.
-func (c *ScrapeConfig) ScrapeInterval() time.Duration {
-	return stringToDuration(c.GetScrapeInterval())
+// UnmarshalYAML implements the yaml.Unmarshaller interface.
+func (c *ScrapeConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	c.DefaultedScrapeConfig = DefaultScrapeConfig
+	err := unmarshal(&c.DefaultedScrapeConfig)
+	if err != nil {
+		return err
+	}
+	if !jobNameRE.MatchString(c.JobName) {
+		return fmt.Errorf("%q is not a valid job name", c.JobName)
+	}
+	return nil
 }
 
-// ScrapeTimeout gets the scrape timeout for the scrape config.
-func (c *ScrapeConfig) ScrapeTimeout() time.Duration {
-	return stringToDuration(c.GetScrapeTimeout())
+// DefaultedScrapeConfig is a proxy type for ScrapeConfig.
+type DefaultedScrapeConfig struct {
+	// The job name to which the job label is set by default.
+	JobName string `yaml:"job_name"`
+	// How frequently to scrape the targets of this scrape config.
+	ScrapeInterval Duration `yaml:"scrape_interval"`
+	// The timeout for scraping targets of this config.
+	ScrapeTimeout Duration `yaml:"scrape_timeout"`
+	// The HTTP resource path on which to fetch metrics from targets.
+	MetricsPath string `yaml:"metrics_path"`
+	// The URL scheme with which to fetch metrics from targets.
+	Scheme string `yaml:"scheme"`
+
+	// List of labeled target groups for this job.
+	TargetGroups []*TargetGroup `yaml:"target_groups,omitempty"`
+	// List of DNS service discovery configurations.
+	DNSConfigs []*DNSConfig `yaml:"dns_configs,omitempty"`
```

Comment:
Just "DNS" sounds ambiguous, since e.g. statically configured targets are also resolved via DNS. The mechanism we're using is usually called DNS-SD (https://www.ietf.org/rfc/rfc6763.txt). Maybe `DNSSDConfigs`, etc.?